File: conformance/tests/p384-wallet/p384-wallet.runar.ts
import { SmartContract, assert, checkSig, hash160, verifyECDSA_P384 } from 'runar-lang';
import type { ByteString, Addr, Sig, PubKey } from 'runar-lang';

class P384Wallet extends SmartContract {
  readonly ecdsaPubKeyHash: Addr;
  readonly p384PubKeyHash: ByteString;

  constructor(ecdsaPubKeyHash: Addr, p384PubKeyHash: ByteString) {
    super(ecdsaPubKeyHash, p384PubKeyHash);
    this.ecdsaPubKeyHash = ecdsaPubKeyHash;
    this.p384PubKeyHash = p384PubKeyHash;
  }

  public spend(p384Sig: ByteString, p384PubKey: ByteString, sig: Sig, pubKey: PubKey) {
    assert(hash160(pubKey) == this.ecdsaPubKeyHash);
    assert(checkSig(sig, pubKey));
    assert(hash160(p384PubKey) == this.p384PubKeyHash);
    assert(verifyECDSA_P384(sig, p384Sig, p384PubKey));
  }
}

File: conformance/tests/p384-wallet/p384-wallet.runar.py
from runar import SmartContract, ByteString, Addr, Sig, PubKey, public, assert_, hash160, check_sig, verify_ecdsa_p384

class P384Wallet(SmartContract):
    ecdsa_pub_key_hash: Addr
    p384_pub_key_hash: ByteString

    def __init__(self, ecdsa_pub_key_hash: Addr, p384_pub_key_hash: ByteString):
        super().__init__(ecdsa_pub_key_hash, p384_pub_key_hash)
        self.ecdsa_pub_key_hash = ecdsa_pub_key_hash
        self.p384_pub_key_hash = p384_pub_key_hash

    @public
    def spend(self, p384_sig: ByteString, p384_pub_key: ByteString, sig: Sig, pub_key: PubKey):
        assert_(hash160(pub_key) == self.ecdsa_pub_key_hash)
        assert_(check_sig(sig, pub_key))
        assert_(hash160(p384_pub_key) == self.p384_pub_key_hash)
        assert_(verify_ecdsa_p384(sig, p384_sig, p384_pub_key))
